### PR TITLE
OODT-945 Rework OODT configuration to make use of Zookeeper for distributed configuration management (GSoC 2017)

### PR DESCRIPTION
**Pull request by IMS94**
This PR includes the Distributed Configuration Management implementation for OODT [OODT-945] along with the distributed configuration publishing CLI [OODT-952].

Explanation on the design of this feature is available at [Rework OODT configuration to make use of Zookeeper for distributed configuration management](https://cwiki.apache.org/confluence/display/OODT/Rework+OODT+configuration+to+make+use+of+Zookeeper+for+distributed+configuration+management).

User documentation on how to use this distributed configuration management feature is available at [OODT Distributed Configuration Management](https://cwiki.apache.org/confluence/display/OODT/OODT+Distributed+Configuration+Management)

Corresponding GSoC project proposal is available at [https://cwiki.apache.org/confluence/display/OODT/Proposal](https://cwiki.apache.org/confluence/display/OODT/Proposal)

**Since this is the first glance of the distributed configuration management feature, there can be bugs and improvements required to be addressed. Please review this PR and provide me with your kind feedback so that I can improve on that.**

Thank you!

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -53,8 +50,7 @@ To contribute a patch, follow these instructions (note that installing
 ```
 
 
-License (see also LICENSE.txt)
-==============================
+## License (see also LICENSE.txt)
```
Comment (chrismattmann) [inline README.md]: great changes, these are orthogonal though - can you separate them into a new PR apart from this?

File: config/src/main/resources/examples/filemgr/cmd-line-actions.xml [hunk 2]
```diff
@@ -0,0 +1,123 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (chrismattmann) [inline config/src/main/resources/examples/filemgr/cmd-line-actions.xml]: maybe we should use maven-ant-tasks to copy these in the prepare-resources phase or in the compile phase (whatever you need them for) instead of physically copying the files?

File: config/src/main/resources/examples/filemgr/filemgr.properties [hunk 3]
```diff
@@ -0,0 +1,114 @@
+#
```
Comment (chrismattmann) [inline config/src/main/resources/examples/filemgr/filemgr.properties]: same why copy these when we can use some maven copy task

File: filemgr/src/main/java/org/apache/oodt/cas/filemgr/system/XmlRpcFileManager.java [hunk 4]
```diff
@@ -52,29 +41,20 @@
...
 import java.net.URI;
 import java.net.URISyntaxException;
 import java.net.URL;
-import java.util.Arrays;
-import java.util.Hashtable;
-import java.util.LinkedList;
-import java.util.List;
-import java.util.Map;
-import java.util.Vector;
+import java.util.*;
```
Comment (chrismattmann) [inline filemgr/src/main/java/org/apache/oodt/cas/filemgr/system/XmlRpcFileManager.java]: we don't use star imports and these changes are orthogonal

**Comment by chrismattmann** [comment]
please see my review comments above - great start!

**Comment by chrismattmann** [comment]
@IMS94

File: README.md (hunk 1 above)
Comment (IMS94) [inline README.md]: @chrismattmann Sure. But I think you have applied these changes to the Readme.md with the commit [update README.md](https://github.com/apache/oodt/commit/1b42cacacfc96e96b69f0c28d7eefcb89ae85856). Therefore, I will rollback this change

File: filemgr/src/main/java/org/apache/oodt/cas/filemgr/system/XmlRpcFileManager.java (hunk 4 above)
Comment (IMS94) [inline filemgr/src/main/java/org/apache/oodt/cas/filemgr/system/XmlRpcFileManager.java]: @chrismattmann Sorry, it was my fault. I was using IntelliJ IDEA's optimize imports command. I will fix this.

File: config/src/main/resources/examples/filemgr/cmd-line-actions.xml (hunk 2 above)
Comment (IMS94) [inline config/src/main/resources/examples/filemgr/cmd-line-actions.xml]: @chrismattmann I will look into that. Thanks for pointing that out. Didn't know we can do that with a plugin before.

Comment (chrismattmann) [inline config/src/main/resources/examples/filemgr/cmd-line-actions.xml]: thanks check out Maven Antrun plugin

File: README.md (hunk 1 above)
Comment (chrismattmann) [inline README.md]: Hi @IMS94 I haven't applied *all of them*, just the *first one*, so can you please open a new PR with the other changes you did and remove/rollback this one.

Comment (IMS94) [inline README.md]: Hi @chrismattmann What I had done on README.md was applying **# Heading 1** (markdown heading styles) for all the headings replacing 

=============================================================
 Welcome to Apache OODT  <http://oodt.apache.org/>
=============================================================
type headings. As I saw, you had fixed the main heading (Apache OODT heading)'s formatting with **# Heading**  style. Now the formatting of README is ok and my previous suggestion is not needed.

Comment (chrismattmann) [inline README.md]: hi @IMS94 I am saying that the rest of your heading formattings (for all else besides the first main heading) are also welcomed, so at a minimum:

1) remove them from this PR
2) if you like (and I am +1 for this so please do) open a new PR with *just those changes for the rest of the headings* and I will gladly commit it. 

HTH.

**Comment by chrismattmann** [comment]
hi @IMS94 I have errors building this locally can you check:

```
LMC-053601:oodt mattmann$ mvn clean install
[INFO] Scanning for projects...
[ERROR] [ERROR] Some problems were encountered while processing the POMs:
[WARNING] 'parent.relativePath' of POM org.apache.oodt:oodt-conf:[unknown-version] (/Users/mattmann/git/oodt/config/pom.xml) points at org.apache.oodt:oodt-core instead of org.apache.oodt:oodt, please verify your project structure @ org.apache.oodt:oodt-conf:[unknown-version], /Users/mattmann/git/oodt/config/pom.xml, line 5, column 13
[ERROR] 'dependencies.dependency.version' for org.apache.curator:curator-framework:jar is missing. @ org.apache.oodt:oodt-conf:[unknown-version], /Users/mattmann/git/oodt/config/pom.xml, line 19, column 21
[ERROR] 'dependencies.dependency.version' for args4j:args4j:jar is missing. @ org.apache.oodt:oodt-conf:[unknown-version], /Users/mattmann/git/oodt/config/pom.xml, line 39, column 21
[ERROR] 'dependencies.dependency.version' for org.apache.curator:curator-test:jar is missing. @ org.apache.oodt:oodt-conf:[unknown-version], /Users/mattmann/git/oodt/config/pom.xml, line 55, column 21
 @ 
[ERROR] The build could not read 1 project -> [Help 1]
[ERROR]   
[ERROR]   The project org.apache.oodt:oodt-conf:1.1-SNAPSHOT (/Users/mattmann/git/oodt/config/pom.xml) has 3 errors
[ERROR]     'dependencies.dependency.version' for org.apache.curator:curator-framework:jar is missing. @ org.apache.oodt:oodt-conf:[unknown-version], /Users/mattmann/git/oodt/config/pom.xml, line 19, column 21
[ERROR]     'dependencies.dependency.version' for args4j:args4j:jar is missing. @ org.apache.oodt:oodt-conf:[unknown-version], /Users/mattmann/git/oodt/config/pom.xml, line 39, column 21
[ERROR]     'dependencies.dependency.version' for org.apache.curator:curator-test:jar is missing. @ org.apache.oodt:oodt-conf:[unknown-version], /Users/mattmann/git/oodt/config/pom.xml, line 55, column 21
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/ProjectBuildingException
```

**Comment by IMS94** [comment]
Hi @chrismattmann, can you check now? I had put wrong parent artifactId name in `config/pom.xml` earlier. It should build now.

**Comment by chrismattmann** [comment]
hi @IMS94 got further, but failing on unit tests, thoughts?

```
-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running org.apache.oodt.config.distributed.DistributedConfigurationManagerTest
log4j:WARN No appenders could be found for logger (org.apache.zookeeper.server.quorum.QuorumPeerConfig).
log4j:WARN Please initialize the log4j system properly.
Starting configuration publishing

Processing commands for component : FILE_MANAGER
Publishing configuration for : FILE_MANAGER
Published configuration for : FILE_MANAGER


Processing commands for component : RESOURCE_MANAGER
Publishing configuration for : RESOURCE_MANAGER
Published configuration for : RESOURCE_MANAGER

Starting configuration publishing

Processing commands for component : FILE_MANAGER
Clearing configuration for : FILE_MANAGER
Cleared configuration for : FILE_MANAGER


Processing commands for component : RESOURCE_MANAGER
Clearing configuration for : RESOURCE_MANAGER
Cleared configuration for : RESOURCE_MANAGER

Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 1.209 sec <<< FAILURE! - in org.apache.oodt.config.distributed.DistributedConfigurationManagerTest
loadConfigurationTest(org.apache.oodt.config.distributed.DistributedConfigurationManagerTest)  Time elapsed: 0.049 sec  <<< FAILURE!
java.lang.AssertionError: null
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertTrue(Assert.java:52)
	at org.apache.oodt.config.distributed.DistributedConfigurationManagerTest.loadConfigurationTest(DistributedConfigurationManagerTest.java:91)

Running org.apache.oodt.config.distributed.DistributedConfigurationPublisherTest
Starting configuration publishing

Processing commands for component : FILE_MANAGER
Publishing configuration for : FILE_MANAGER
Published configuration for : FILE_MANAGER

Verifying configuration for : FILE_MANAGER
OK... Configuration verified
Verified configuration for : FILE_MANAGER


Processing commands for component : RESOURCE_MANAGER
Publishing configuration for : RESOURCE_MANAGER
Published configuration for : RESOURCE_MANAGER

Verifying configuration for : RESOURCE_MANAGER
OK... Configuration verified
Verified configuration for : RESOURCE_MANAGER

Starting configuration publishing

Processing commands for component : FILE_MANAGER
Clearing configuration for : FILE_MANAGER
Cleared configuration for : FILE_MANAGER


Processing commands for component : RESOURCE_MANAGER
Clearing configuration for : RESOURCE_MANAGER
Cleared configuration for : RESOURCE_MANAGER

Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.512 sec - in org.apache.oodt.config.distributed.DistributedConfigurationPublisherTest

Results :

Failed tests: 
  DistributedConfigurationManagerTest.loadConfigurationTest:91 null

Tests run: 2, Failures: 1, Errors: 0, Skipped: 0

[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] OODT Core .......................................... SUCCESS [  1.820 s]
[INFO] Common Utilities ................................... SUCCESS [ 17.797 s]
[INFO] CAS Command Line Interface ......................... SUCCESS [  6.046 s]
[INFO] OODT - Configuration Management .................... FAILURE [  5.426 s]
[INFO] Process Control System Input Data Package .......... SKIPPED
[INFO] Catalog and Archive Service Generic Multi-valued Metadata Container SKIPPED
[INFO] CAS Protocol ....................................... SKIPPED
[INFO] CAS Protocol FTP Implementation .................... SKIPPED
[INFO] CAS Protocol HTTP Implementation ................... SKIPPED
[INFO] CAS Protocol IMAPS Implementation .................. SKIPPED
[INFO] CAS Protocol SFTP Implementation ................... SKIPPED
[INFO] Query Expression ................................... SKIPPED
[INFO] OODT Single Sign On Security Package ............... SKIPPED
[INFO] Catalog and Archive File Management Component ...... SKIPPED
[INFO] OODT CAS Virtual Catalog and Integration Service. .. SKIPPED
[INFO] Catalog and Archive Resource Management Component .. SKIPPED
[INFO] Catalog and Archive Workflow Management Component .. SKIPPED
[INFO] Catalog and Archive Crawling Framework ............. SKIPPED
[INFO] CAS Curation Web Services .......................... SKIPPED
[INFO] Process Control System Core Package ................ SKIPPED
[INFO] OODT Wicket Web Components ......................... SKIPPED
[INFO] CAS Curation Interface ............................. SKIPPED
[INFO] CAS PGE Adaptor Framework .......................... SKIPPED
[INFO] CAS Installer Maven Mojo ........................... SKIPPED
[INFO] OODT :: Archetypes :: OpsUI ........................ SKIPPED
[INFO] OODT :: Archetypes :: RADiX ........................ SKIPPED
[INFO] OODT :: Archetypes ................................. SKIPPED
[INFO] CAS Push-Pull-Framework ............................ SKIPPED
[INFO] Product Service .................................... SKIPPED
[INFO] Profile Service .................................... SKIPPED
[INFO] OODT Web Grid ...................................... SKIPPED
[INFO] XML-configured, DBMS-based Product and Profile Server SKIPPED
[INFO] Apache OODT Configurable OPeNDAP Profile Server .... SKIPPED
[INFO] CAS File Manager Browser Web App ................... SKIPPED
[INFO] CAS Product Server Web Application ................. SKIPPED
[INFO] CAS Workflow Manager Monitor Web App ............... SKIPPED
[INFO] Catalog and Archive File Management Browser ........ SKIPPED
[INFO] Catalog and Archive Workflow Management GUI Editor . SKIPPED
[INFO] Process Control System Operator Interface Webapp ... SKIPPED
[INFO] OODT Process Control System JAX-RS service layer ... SKIPPED
[INFO] Apache OODT ........................................ SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 31.600 s
[INFO] Finished at: 2017-07-13T15:38:14-07:00
[INFO] Final Memory: 53M/480M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.17:test (default-test) on project oodt-conf: There are test failures.
[ERROR] 
[ERROR] Please refer to /Users/mattmann/git/oodt/config/target/surefire-reports for the individual test results.
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :oodt-conf
LMC-053601:oodt mattmann$ 
```

**Comment by IMS94** [comment]
Hi @chrismattmann, Can let me know do you have any of previously set up `FILEMGR_HOME` or `RESMGR_HOME` environment variables available in your system?

**Comment by chrismattmann** [comment]
hi @IMS94 do I need to set those env vars? This will screw up tests since they should not have environment variable requirements, or if so, should be mock'ed in the tests themselves.

**Comment by chrismattmann** [comment]
So I checked, and I do have those environment variables set, except for ```RESMGR_HOME``` e.g., 

```
LMC-053601:oodt mattmann$ echo $FILEMGR_HOME
/usr/local/bigtranslate/filemgr
LMC-053601:oodt mattmann$ echo $FILEMGR_URL
http://localhost:9000
LMC-053601:oodt mattmann$ 
LMC-053601:oodt mattmann$ echo $RESMGR_URL
http://localhost:9002
LMC-053601:oodt mattmann$ 
```

I was thinking about your code - is it that if an environment variable from local config isn't that set - you are publishing a null value to ZK or there is a getPropertyOrNull call somewhere that is causing this? If so, we should probably have a sensibleDefaults() configuration which are all the sensible defaults in case the environment is Null. Even better would be to use PathUtils in o.a.cas.metadata.util to do EnvVarOrMetadata replacement on all paths you read from config.

**Comment by IMS94** [comment]
Hi @chrismattmann, No. Tests do not require any environment variable to be set. But if any of the above mentioned env variables are set, that could be a possible reason for your test failure (because I didn't encounter test failures). I'm using environment variables to determine each OODT component's HOME when downloading and storing configuration files from zookeeper.

Therefore, what I wanted know was whether you have any of those env vars set.

**Comment by IMS94** [comment]
Hi @chrismattmann, Thanks for the update. That is the reason for the test failure. Actually this is a case I should have addressed within my code. I have wrote the fix. Will push it now.

**Comment by chrismattmann** [comment]
@IMS94 feel free to request an invite to uscdatascience.slack.com I'll make a channel for slack chat if you want real time comm on this

**Comment by chrismattmann** [comment]
OK @IMS94 I have invited you to the USC Data Science slack

**Comment by IMS94** [comment]
Hi @chrismattmann, Thanks for the invitation. 

I have pushed an update. There's still a chance for the tests to fail because I store configuration files in `${COMPONENT_HOME}` directory if `${COMPONENT_HOME}` is present, else in the working directory. If it fails again, I will find a more convenient method for my implementation.

**Comment by chrismattmann** [comment]
merged this thanks @IMS94 

```
LMC-053601:oodt mattmann$ git push -u origin feature/zookeeper-config
Counting objects: 475, done.
Delta compression using up to 8 threads.
Compressing objects: 100% (203/203), done.
Writing objects: 100% (475/475), 81.85 KiB | 0 bytes/s, done.
Total 475 (delta 155), reused 355 (delta 113)
remote: oodt git commit: Update log4j config.
remote: oodt git commit: Merge branch 'feature/zookeeper-config' of https://github.com/IMS94/oodt into feature/zookeeper-config
remote: oodt git commit: assert not null.
remote: oodt git commit: Fix for empty component home variable
remote: oodt git commit: Merge and apply robust changes for stack trace print in tests.
remote: oodt git commit: Merge branch 'feature/zookeeper-config' of https://github.com/IMS94/oodt into feature/zookeeper-config
remote: oodt git commit: fix for presence of component home environment variables in tests
remote: oodt git commit: Merge branch 'feature/zookeeper-config' of https://github.com/IMS94/oodt into feature/zookeeper-config
remote: oodt git commit: fix pom.xml
remote: oodt git commit: Fixed parent artifactId in config/pom.xml
remote: oodt git commit: Removed * imports. Improved code formatting and refactored code.
remote: oodt git commit: Merge branch 'master' of https://github.com/apache/oodt into feature/zookeeper-config
remote: oodt git commit: Update README.md
remote: oodt git commit: Merge branch 'OODT-847' of http://github.com/adhulipa/oodt
remote: oodt git commit: Record change for OODT-847.
remote: oodt git commit: Merge branch 'master' into feature/zookeeper-config
remote: oodt git commit: Improvements to file downloading from zookeeper and tests
remote: oodt git commit: Minor improvements for configuration manager tests
remote: oodt git commit: Created test cases for distributed configuration manager and publisher
remote: oodt git commit: Added zip packaging for distributed configuration publisher
remote: oodt git commit: Implemented configuration fetching for DistributedConfigurationManager
remote: oodt git commit: Altered configuration publishing mechanism to use spring configuration
remote: oodt git commit: Implemented global configuration publishing mechanism.
remote: oodt git commit: Defined initial API for ConfigurationManager
remote: oodt git commit: Fix for OODT-847 contributed by adhulipa & rverma
To https://git-wip-us.apache.org/repos/asf/oodt.git
   ec5b201..3fb7660  feature/zookeeper-config -> feature/zookeeper-config
Branch feature/zookeeper-config set up to track remote branch feature/zookeeper-config from origin.
LMC-053601:oodt mattmann$ 
```

**Comment by IMS94** [comment]
@chrismattmann Thanks for your support too. Did the tests pass for you without a problem?

**Comment by chrismattmann** [comment]
they did pass once I disabled tesseract
